pkg/operator/ceph/cluster/dependents.go

```diff
@@ -0,0 +1,146 @@
+		"CephNFSes",
+		"CephClients",
+	}
+)
+
+// CephClusterDependents returns a DependentList of dependents of a CephCluster in the namespace.
+func CephClusterDependents(c *clusterd.Context, namespace string) (*DependentList, error) {
+	ctx := context.TODO()
+
+	dependents := NewDependentList()
+	errs := []error{}
+
+	for _, pluralKind := range cephClusterDependentPluralKinds {
+		resource := pluralKindToResource(pluralKind)
+		gvr := cephv1.SchemeGroupVersion.WithResource(resource)
+		list, err := c.DynamicClientset.Resource(gvr).Namespace(namespace).List(ctx, metav1.ListOptions{})
+		if err != nil {
+			errs = append(errs, errors.Wrapf(err, "failed to list %s", pluralKind))
+			continue
+		}
+		if len(list.Items) > 0 {
+			for _, obj := range list.Items {
+				dependents.Add(pluralKind, obj.GetName())
+			}
+		}
+	}
+	// returns a nil error if there are no errors in the list
+	outErr := util.AggregateErrors(errs, "failed to list some dependents for CephCluster in namespace %q", namespace)
+
+	return dependents, outErr
+}
+
+func pluralKindToResource(pluralKind string) string {
+	// The dynamic client wants resources which are lower-case versions of the plural Kinds for
+	// Kubernetes CRDs in almost all cases.
+	return strings.ToLower(pluralKind)
+}
+
+// A DependentList represents a list of dependents of a resource. Each dependent has a Kind and a
+// list of names of dependent resources.
+type DependentList struct {
+	d map[string][]string // map from dependent Resource to a list of dependent names
+}
+
+// NewDependentList creates a new empty DependentList.
+func NewDependentList() *DependentList {
+	return &DependentList{
+		d: make(map[string][]string),
+	}
+}
+
+// Empty returns true if the DependentList is empty or false otherwise.
+func (d *DependentList) Empty() bool {
+	return len(d.d) == 0
+}
+
+// Add adds a dependent name for a plural Kind to the DependentList.
+func (d *DependentList) Add(pluralKind string, name string) {
+	names, ok := d.d[pluralKind]
+	if !ok {
+		d.d[pluralKind] = []string{name}
+		return
+	}
+	d.d[pluralKind] = append(names, name)
+}
+
+func (d *DependentList) PluralKinds() []string {
+	kinds := []string{}
+	for k := range d.d {
+		kinds = append(kinds, k)
+	}
+	return kinds
+}
+
+// OfPluralKind returns the names of dependents of the Kind (plural), or an empty list if no
+// dependents exist.
+func (d *DependentList) OfPluralKind(pluralKind string) []string {
+	names, ok := d.d[pluralKind]
+	if !ok {
+		return []string{}
+	}
+	return names
+}
+
+// StringWithHeader outputs the dependent list as a pretty-printed string headed with the given
+// formatting directive (followed by a colon). It outputs dependents in alphabetical order by the
+// plural Kind.
+// Example:
+//    StringWithHeader("dependents of my %q", "mom")  -->
+//    `dependents of my "mom": FirstResources: [name1], SecondResources: [name2 name2 name3]`
+func (d *DependentList) StringWithHeader(headerFormat string, args ...interface{}) string {
```

Comment:
This would work with any slice, right? Could it be moved to the `pkg/util` package?

Reply:
It does special processing to look through the map-string-to-slice-of-strings, so I don't think it's particularly reusable.

Reply:
if not used outside this package it could be private
```suggestion
func (d *DependentList) stringWithHeader(headerFormat string, args ...interface{}) string {
```

Reply:
This will soon be made into a non-private package with future work on dependencies. This will get reused.

Reply:
Sure, how about moving to that package in this PR?